show current config and where the config values were set at -vvv and higher

alikins opened the pull request:
##### SUMMARY
At verbosity levels of '-vvv', display a summary of the current config including info about where that config value came from. Output the format in an 'ini' format.

ie, like
``` ini
[defaults]
# from file ./ansible_min.cfg
bin_ansible_callbacks = true

# from file ./ansible_min.cfg
forks = 1

# from file ./ansible_min.cfg
roles_path = $HOME/ansible/galaxy-roles:$HOME/ansible/roles:/etc/ansible/roles

# from file ./ansible_min.cfg
nocows = 1

# from env var ANSIBLE_FILTER_PLUGINS
filter_plugins = /home/adrian/whatever_plugins

# from file ./ansible_min.cfg
action_plugins = $HOME/ansible/my-plugins/action

# from file ./ansible_min.cfg
callback_plugins = $HOME/ansible/my-plugins/callback

[diff]
# from env var ANSIBLE_DIFF_ALWAYS
always = 1

[ssh_connection]
# from env var ANSIBLE_SSH_RETRIES
retries = 2

# from file ./ansible_min.cfg
pipelining = False

# from file ./ansible_min.cfg
ssh_args = -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=600s -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r -o ConnectTimeout=2s
```



##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - Bugfix Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/constants.py
lib/ansible/cli/__init__.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
[fedora-25:ansible (dump_config_source %)]$ ansible --version
ansible 2.3.0 (dump_config_source 71a3d91b9c) last updated 2017/03/09 15:58:12 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan 12 2017, 17:59:37) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]


```


##### ADDITIONAL INFORMATION
example
```
[fedora-25:ansible (dump_config_source %)]$ cat ansible_min.cfg 
[defaults]
roles_path = $HOME/ansible/galaxy-roles:$HOME/ansible/roles:/etc/ansible/roles
forks = 1
nocows = 1
bin_ansible_callbacks = true
callback_plugins   = $HOME/ansible/my-plugins/callback
action_plugins = $HOME/ansible/my-plugins/action
filter_plugins = $HOME/ansible/my-plugins/filter

[ssh_connection]
pipelining = False
ssh_args = -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=600s -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r -o ConnectTimeout=2s

[paramiko]
[fedora-25:ansible (dump_config_source %)]$ env| grep ^ANSIBLE
ANSIBLE_DIFF_ALWAYS=1
ANSIBLE_SRC_DIR=/home/adrian/src/ansible
ANSIBLE_HOME=/home/adrian/src/ansible
ANSIBLE_MODULES_CORE_SRC_DIR=/home/adrian/src/ansible-modules-core
[fedora-25:ansible (dump_config_source %)]$ ANSIBLE_CONFIG=./ansible_min.cfg ANSIBLE_FILTER_PLUGINS=~/whatever_plugins ANSIBLE_SSH_RETRIES=2 ansible -vvv localhost -m ping
Using ./ansible_min.cfg as config file
Current config:
[defaults]
# from file ./ansible_min.cfg
bin_ansible_callbacks = true

# from file ./ansible_min.cfg
forks = 1

# from file ./ansible_min.cfg
roles_path = $HOME/ansible/galaxy-roles:$HOME/ansible/roles:/etc/ansible/roles

# from file ./ansible_min.cfg
nocows = 1

# from env var ANSIBLE_FILTER_PLUGINS
filter_plugins = /home/adrian/whatever_plugins

# from file ./ansible_min.cfg
action_plugins = $HOME/ansible/my-plugins/action

# from file ./ansible_min.cfg
callback_plugins = $HOME/ansible/my-plugins/callback

[diff]
# from env var ANSIBLE_DIFF_ALWAYS
always = 1

[ssh_connection]
# from env var ANSIBLE_SSH_RETRIES
retries = 2

# from file ./ansible_min.cfg
pipelining = False

# from file ./ansible_min.cfg
ssh_args = -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=600s -o ControlPath=/home/adrian/.ansible/cp/ansible-ssh-%h-%p-%r -o ConnectTimeout=2s

 [WARNING]: provided hosts list is empty, only localhost is available

META: ran handlers
Using module file /home/adrian/src/ansible/lib/ansible/modules/system/ping.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: adrian
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/adrian/.ansible/tmp/ansible-tmp-1489093400.91-228155790759657 `" && echo ansible-tmp-1489093400.91-228155790759657="` echo /home/adrian/.ansible/tmp/ansible-tmp-1489093400.91-228155790759657 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmp79jutB TO /home/adrian/.ansible/tmp/ansible-tmp-1489093400.91-228155790759657/ping.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/adrian/.ansible/tmp/ansible-tmp-1489093400.91-228155790759657/ /home/adrian/.ansible/tmp/ansible-tmp-1489093400.91-228155790759657/ping.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/adrian/.ansible/tmp/ansible-tmp-1489093400.91-228155790759657/ping.py; rm -rf "/home/adrian/.ansible/tmp/ansible-tmp-1489093400.91-228155790759657/" > /dev/null 2>&1 && sleep 0'
localhost | SUCCESS => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "data": null
        }
    }, 
    "ping": "pong"
}
META: ran handlers
META: ran handlers
```

<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

alikins committed:
Add a way to dump config info at -vvv or higher

add a dumps_config() to constants.py

Show the config if we are running at -vvv or higher.

bcoca commented:
part of the ansible-config proposal

https://github.com/ansible/proposals/issues/35
https://github.com/ansible/ansible/pull/12797

but expanded to 'merged' config files and add host/group vars when 'connection variables' are invovled

alikins committed:
skip (finding and then ignoring) defaults for now

alikins committed:
skip 'comment' for now as well

alikins committed:
show the default value if it is interesting

interesting here meaning, not None

alikins committed:
sort the output by section, then by config key

Since the dicts are unordered, sort the sections and keys
so they are written out in a stable order.

alikins commented:
Example with the 'interesting defaults' and sorting of sections and keys:

```
[fedora-25:ansible (dump_config_source %)]$ ANSIBLE_CONFIG=~/ansible/ansible.cfg ANSIBLE_FILTERS=~/foobar ANSIBLE_CALLBACK_PLUGINS=~/foo/callback/ ansible -vvv localhost -m ping
Using /home/adrian/ansible/ansible.cfg as config file
Current config:
[defaults]
# from file /home/adrian/ansible/ansible.cfg
# default: ~/.ansible/plugins/action:/usr/share/ansible/plugins/action
action_plugins = $HOME/ansible/my-plugins/action

# from file /home/adrian/ansible/ansible.cfg
# default: False
bin_ansible_callbacks = true

# from environment variable ANSIBLE_CALLBACK_PLUGINS
# default: ~/.ansible/plugins/callback:/usr/share/ansible/plugins/callback
callback_plugins = /home/adrian/foo/callback/

# from file /home/adrian/ansible/ansible.cfg
# default: ~/.ansible/plugins/filter:/usr/share/ansible/plugins/filter
filter_plugins = $HOME/ansible/my-plugins/filter

# from file /home/adrian/ansible/ansible.cfg
# default: 5
forks = 1

# from file /home/adrian/ansible/ansible.cfg
nocows = 1

# from file /home/adrian/ansible/ansible.cfg
# default: /etc/ansible/roles
roles_path = $HOME/ansible/galaxy-roles:$HOME/ansible/roles:/etc/ansible/roles

[diff]
# from environment variable ANSIBLE_DIFF_ALWAYS
# default: False
always = 

[ssh_connection]
# from file /home/adrian/ansible/ansible.cfg
# default: False
pipelining = False
```
